app/Utils/DatabaseCleanupUtils.php

```diff
@@ -366,4 +366,49 @@ private static function deleteRowsChunked(string $query, array $ids): void
             usleep(1);
         }
     }
+
+    /** Delete unused rows in batches */
+    public static function deleteUnusedRows(string $table, string $field, string $targettable, string $selectfield = 'id'): void
+    {
+        $start = DB::table($table)->min($field);
+        $max = DB::table($table)->max($field);
+        if (!is_numeric($start) || !is_numeric($max)) {
+            Log::info("Could not determine min and max for `{$field}` on `{$table}`");
+            return;
+        }
+
+        $start = intval($start);
+        $max = intval($max);
+
+        $total = $max - $start;
+        if ($total < 1) {
+            return;
+        }
+        $num_done = 0;
+        $num_deleted = 0;
+        $next_report = 10;
+        $done = false;
+        Log::info("Deleting unused rows from `{$table}`");
+        while (!$done) {
+            $end = $start + 49999;
+            $num_deleted += DB::delete("
+                DELETE FROM $table
+                WHERE $field BETWEEN $start AND $end
+                      AND $field NOT IN (SELECT $selectfield FROM $targettable)");
+            $num_done += 50000;
+            if ($end >= $max) {
+                $done = true;
+            } else {
+                usleep(1);
+                $start += 50000;
+                // Calculate percentage of work completed so far.
+                $percent = round(($num_done / $total) * 100, -1);
+                if ($percent > $next_report) {
+                    Log::info("{$next_report}%");
```

Comment:
It would be good to add some text before the percentage to indicate which job it comes from in case other log messages get interleaved with these.